src/api/app/models/concerns/staging_project.rb

```diff
@@ -142,35 +142,21 @@ def cache_problems
   end
 
   def state
-    return :empty unless staged_requests.exists?
-    return :unacceptable if unacceptable_state?
-    return :review if review_state?
-    general_state
-  end
-
-  def unacceptable_state?
-    untracked_requests.present? || staged_requests.obsolete.present?
-  end
-
-  def review_state?
-    general_state == :acceptable && missing_reviews.present?
-  end
-
-  def general_state
-    build_state == :acceptable ? check_state : build_state
+    return :empty if staged_requests.blank?
+    return :unacceptable if untracked_requests.present? || staged_requests.obsolete.exists?
+    bc_state = build_or_check_state
+    return bc_state if bc_state
+    return :review if missing_reviews.present?
+    :acceptable
   end
 
-  def build_state
+  def build_or_check_state
+    # build_state
     return :building if building_repositories.present? || disabled_repositories.present?
     return :failed if broken_packages.present?
-
-    :acceptable
-  end
-
-  def check_state
+    # check_state
     return :testing if missing_checks.present? || checks.pending.exists?
-    return :failed if checks.failed.exists?
-    return :acceptable
+    :failed if checks.failed.exists?
```

Comment:
Ok, I have done a mix of both suggestions. I created a single method called `build_or_check_state` and I also assigned it to a variable avoiding to call it twice.

@Ana06, @hennevogel can you have another look?

PS: I guess it's not a good idea to change the order of the lines, for this reason I don't join the two `failed`cases like:

```                              
    :failed if broken_packages.present? || checks.failed.exists?                                         
```

What do you think?